sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +203,216 @@ object Hash {
+    ): NodeBuilder = {
+      NodeBuilder.builderForNode(node, hashTracer).addVersion.addNode(node, nodes)
+    }
+
+    @throws[NodeHashingError]
+    protected def addNodeFromNodeId(nodes: Map[NodeId, Node]): (this.type, NodeId) => this.type =
+      (builder, nodeId) => {
+        val node = nodes.getOrElse(nodeId, throw NodeHashingError.IncompleteTransactionTree(nodeId))
+        // For inner node with add the hashed value of the node to the current builder. We do not inline
+        // the encoded node. This is to keep NodeBuilder self contained w.r.t to its message digest.
+        addHash(
+          builder.addFromVersionedNode(node, nodes, hashTracer.subNodeTracer).build,
+          "(Hashed Inner Node)",
+        )
+      }
+
+    def addNodesFromNodeIds(nodeIds: ImmArray[NodeId], nodes: Map[NodeId, Node]): this.type =
+      iterateOver(nodeIds)(addNodeFromNodeId(nodes))
+
+    def addTransactionVersion(transactionVersion: TransactionVersion): this.type = {
+      addString(TransactionVersion.toProtoValue(transactionVersion))
+    }
+  }
+
+  /** Class with additional methods to hash nodes. Uses a single MessageDigest to hash the entire node including all its values.
+    */
+  private sealed abstract class NodeBuilder(
+      version: NodeHashVersion,
+      hashTracer: HashTracer,
+  ) extends TransactionBuilder(hashTracer) {
+
+    override def addVersion: this.type = {
+      super.addVersion
+        .addByte(version.id, s"${formatByteToHexString(version.id)} (Node Encoding Version)")
+    }
+
+    def addNode(node: Node, nodes: Map[NodeId, Node]): NodeBuilder
+  }
+
+  private object NodeBuilder {
+    private[lf] val LanguageVersionToNodeBuilder: Map[LanguageVersion, HashTracer => NodeBuilder] =
+      Map(
+        LanguageVersion.v2_1 -> (hashTracer => new NodeBuilderV1(hashTracer))
+      )
+
+    private[crypto] sealed abstract class NodeTag(val tag: Byte)
+    private[crypto] object NodeTag {
+      case object CreateTag extends NodeTag(0)
+      case object ExerciseTag extends NodeTag(1)
+      case object FetchTag extends NodeTag(2)
+      case object LookupTag extends NodeTag(3)
+      case object RollbackTag extends NodeTag(4)
+    }
+
+    @throws[NodeHashingError]
+    private[lf] def builderForVersion(
+        version: LanguageVersion,
+        hashTracer: HashTracer,
+    ): NodeBuilder =
+      NodeBuilder.LanguageVersionToNodeBuilder
+        .getOrElse(version, throw NodeHashingError.UnsupportedLanguageVersion(version))
+        .apply(hashTracer)
+
+    private[lf] def builderForNode(
+        node: Node,
+        hashTracer: HashTracer,
+    ): NodeBuilder = node.optVersion match {
+      case Some(version) => builderForVersion(version, hashTracer)
+      // If the node is version agnostic, the builder doesn't matter so pick V1
+      // This only applies to rollback nodes
+      case None => new NodeBuilderV1(hashTracer)
+    }
+  }
+
+  private final class NodeBuilderV1(hashTracer: HashTracer)
```

Comment:
minor: move the classes for your v1 hashing scheme to a separate file to make the versioning clear?